Improve thread-safety of lazy initializations

scott-ferguson-unity opened the pull request:
Fixes issues with thread safety on the lazy initializations in Mono.Cecil.

The `Interlocked.Exchange` calls will ensure that the same object is returned from each call even under a race (although we my create more than once, but they are just empty collections).  It should also make sure that no write re-ordering issues occur on platforms that allow that.

The locks on `Module.SyncRoot` are a pattern I copied from elsewhere in the code base.  They are always around a call to `Module.Read` which always takes that same lock itself, so no new locking is really being introduced.

scott-ferguson-unity committed:
Improve thread-safety of lazy initializations

mrvoorhe reviewed (approved):
Looks good!

mrvoorhe reviewed (approved):
Looks good!

joncham commented on Mono.Cecil.Cil/Symbols.cs:
```diff
@@ -408,7 +429,14 @@ public sealed class ImportDebugInformation : DebugInformation {
-			get { return targets ?? (targets = new Collection<ImportTarget> ()); }
+			get
+			{
+				if (targets != null)
```
Nit, is there a reason this pattern is inverted from the others?

scott-ferguson-unity replied:
Nope, updated

joncham commented on Mono.Cecil/FieldDefinition.cs:
```diff
@@ -37,7 +38,11 @@ void ResolveLayout ()
 
-			offset = Module.Read (this, (field, reader) => reader.ReadFieldLayout (field));
+			lock (Module.SyncRoot) {
+				if (offset != Mixin.NotResolvedMarker)
```
Why the additional `if` check?

joncham replied:
Never mind, I see the double checked locking pattern now.

joncham commented:
Have you opened equivalent PR to upstream Cecil for comment?

scott-ferguson-unity committed:
Updated lazy initialization for AssemblyNameReference.PublicKeyToken

Intialize it to null to match pattern for the other references.
Fixes the case were the lazy initialization would break when
public_key_token was Empty<byte>.Array.
